### PR TITLE
BigQuery table description

### LINKED ISSUE #406: Support description for BigQuery table
We already support description for fields, we should add support for description of tables too.

Ideally it should just reuse the description annotation but on the class like:

```
@BigQueryType.toTable
@description("table description foodbar") 
case class User(@description("user name") name: String,
                @description("user age") age: Int)
```

### PR DESCRIPTION
**Pull request by ravwojdyla**
Allows to:

```
@BigQueryType.toTable
@description("table description foodbar") 
case class User(@description("user name") name: String,
                @description("user age") age: Int)
```

- [x] get feedback about WIP
- [x] add support for toTable
- [x] move upstream PR to Beam
- [x] move scio PR to apache-beam branch
- [x] add support for fromSchema
- [x] add tests
- [x] clean up code

This includes code/tests from GoogleCloudPlatform/DataflowJavaSDK/pull/535

Fixes #406

### CODE REVIEW

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala [hunk 1]
```diff
@@ -79,14 +79,21 @@ private[types] object TypeProvider {
     checkMacroEnclosed(c)
 
     val (r, caseClassTree, name) = annottees.map(_.tree) match {
-      case List(q"case class $name(..$fields) { ..$body }") =>
+      case l @ List(q"$mods class $name[..$_] $_(..$fields) extends { ..$_ } with ..$_ { $_ => ..$body }") if mods.asInstanceOf[Modifiers].hasFlag(Flag.CASE) =>
+
+        val desc = l.head.asInstanceOf[ClassDef].mods.annotations
+          .filter(_.children.head.toString() == "new description")
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: Can't seem to find a better way to get this - `tpe` for each child is `null` :( any ideas here?

**Comment by ravwojdyla** [comment]
@nevillelyh @andrewsmartin initial -1/+1, should I continue on with the `fromSchema`?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala [hunk 2]
```diff
@@ -79,14 +79,25 @@ private[types] object TypeProvider {
     checkMacroEnclosed(c)
 
     val (r, caseClassTree, name) = annottees.map(_.tree) match {
-      case List(q"case class $name(..$fields) { ..$body }") =>
+      case l @ List(q"$mods class $name[..$tparams] $ctorMods(..$fields) extends { ..$earlydefns } with ..$parents { $self => ..$body }") if mods.asInstanceOf[Modifiers].hasFlag(Flag.CASE) =>
```
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: Are all of these necessary? It's just the `@description` annotation you need right? Do we want to allow other traits on `toTable` classes?

**Comment by nevillelyh** [comment]
Maybe better to work on the `apache-beam` branch directly and avoid cherry-picking / diverging monkey patched classes?
Do you need to submit PR to Beam if we go that route?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala (hunk 2 above)
Comment (andrewsmartin) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: Actually just noticed below you're validating against this.

**Comment by andrewsmartin** [comment]
👍 LGTM

**Comment by ravwojdyla** [comment]
@nevillelyh i mean this is how we have done it so far, and it worked "fine". Not sure how long it will take for beam/dataflow-sdk to merge it, it's not very responsive process.

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala (hunk 2 above)
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: The problem here was that all the other shorter quasiquotes I have tried were not accepting extra annotations as `mods`. Only this full quote gave me the annotations in the `mods`, if there is better way - please let me know.

**Comment by nevillelyh** [comment]
Yeah I'm fine with monkey patched files but it'll make my life easier if it's in `apache-beam` branch only so I don't have to cherry pick. This is not a critical feature so we don't need it to be available ASAP in Scio 0.2.x right?

**Comment by ravwojdyla** [comment]
@andrewsmartin @nevillelyh - i might need to rebase, but otherwise ready for review here.

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala [hunk 3]
```diff
@@ -66,19 +66,33 @@ private[types] object TypeProvider {
     schemaToType(c)(schema, annottees, traits, overrides)
   }
 
+  private def getTableDescription(c: blackbox.Context)(tree: Seq[c.universe.Tree])
+  : List[c.universe.Tree] = {
+    import c.universe._
+    tree.head.asInstanceOf[ClassDef].mods.annotations
+      .filter(_.children.head.toString() == "new description")
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: The `tpe` gives me nulls, is there a better way to do this?

File: scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala [hunk 4]
```diff
@@ -142,13 +156,23 @@ private[types] object TypeProvider {
     val (fields, records) = toFields(schema.getFields)
 
     val (r, caseClassTree, name) = annottees.map(_.tree) match {
-      case List(q"class $name") =>
+      case l @ List(q"$mods class $name[..$tparams] $ctorMods(..$cfields) extends { ..$earlydefns } with ..$parents { $self => ..$body }") if mods.asInstanceOf[Modifiers].flags == NoFlags =>
+        if (parents.map(_.toString()).toSet != Set("scala.AnyRef")) {
```
Comment (ravwojdyla) [inline scio-bigquery/src/main/scala/com/spotify/scio/bigquery/types/TypeProvider.scala]: Again the parents `tpe` give me `null`s here, do you know a better way?

**Comment by ravwojdyla** [comment]
fyi BEAM-1301 was merged into beam code base.

**Comment by ravwojdyla** [comment]
Let's wait till the upstream table description is released. Closing for now.

**Pull request closed without merging**